flang/lib/Optimizer/Transforms/OpenACCDataOperandConversion.cpp

```diff
@@ -0,0 +1,198 @@
+#include "mlir/Transforms/DialectConversion.h"
+
+#define DEBUG_TYPE "flang-openacc-conversion"
+#include "../CodeGen/TypeConverter.h"
+
+using namespace mlir;
+
+//===----------------------------------------------------------------------===//
+// Conversion patterns
+//===----------------------------------------------------------------------===//
+
+namespace {
+
+template <typename Op>
+class LegalizeDataOpForLLVMTranslation : public OpConversionPattern<Op> {
+public:
+  explicit LegalizeDataOpForLLVMTranslation(TypeConverter &converter,
+                                            MLIRContext *ctx)
+      : OpConversionPattern<Op>(converter, ctx) {}
+
+  mlir::LogicalResult
+  matchAndRewrite(Op op, llvm::ArrayRef<mlir::Value> operands,
+                  ConversionPatternRewriter &builder) const override {
+    Location loc = op.getLoc();
+    fir::LLVMTypeConverter &converter = *static_cast<fir::LLVMTypeConverter *>(
+        OpConversionPattern<Op>::getTypeConverter());
+    unsigned numDataOperand = op.getNumDataOperands();
+
+    // Keep the non data operands without modification.
+    auto nonDataOperands =
+        operands.take_front(operands.size() - numDataOperand);
+    llvm::SmallVector<mlir::Value> convertedOperands;
+    convertedOperands.append(nonDataOperands.begin(), nonDataOperands.end());
+
+    // Go over the data operand and legalize them for translation.
+    for (unsigned idx = 0; idx < numDataOperand; ++idx) {
+      mlir::Value originalDataOperand = op.getDataOperand(idx);
+
+      //
+      if (auto refTy =
+              originalDataOperand.getType().dyn_cast<fir::ReferenceType>()) {
+        if (refTy.getEleTy().isa<fir::SequenceType>() ||
+            fir::isa_std_type(refTy.getEleTy())) {
+          // Basic arrays and scalars are passed as llvm.ptr to the translation.
+          // Code to compute the size will be generated during the translation.
+          auto convertedType = converter.convertType(refTy);
+          auto convertedValue = builder.create<fir::CastOp>(
+              loc, convertedType, originalDataOperand);
+          convertedOperands.push_back(convertedValue);
+        } else if (auto boxTy = refTy.getEleTy().dyn_cast<fir::BoxType>()) {
+          // BoxType needs more work to extract the correct information.
+          return builder.notifyMatchFailure(
+              op, "fir.box type currently not supported");
+
+          // When a BoxType is encountered we need to push the descriptor and
+          // the actual data.
+
+          auto convertedType =
+              converter.convertType(refTy).cast<mlir::LLVM::LLVMPointerType>();
+          auto convertedValue =
+              builder
+                  .create<fir::CastOp>(loc, convertedType, originalDataOperand)
+                  .res();
+
+          // TODO: currently this code is not doing what it should just so dummy
+          // code to create and populate the data descriptor with info.
+          auto fortranDescriptor =
```

Comment:
The current code is doesn't make much sense but it's just here as a placeholder for when the correct value are extracted or computed. 

Can we compute the size of the descriptor with the ptr diff trick?

Do we have a runtime function to get the size of the data hold be the descriptor?

Reply:
Thanks for the answers. I'll look into that.